### PR TITLE
Add section on automation via workflows.

### PR DESCRIPTION
**Pull request by msporny**
This section adds a section on issuance automation via workflows. This is what creates the ability to do things like AutoRefresh2021. If folks are ok with the non-normative text here (the concept), then I plan to add normative text in a separate PR to try and establish something that is testable.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c-ccg/vc-api/pull/254.html" title="Last updated on Jan 7, 2022, 4:53 PM UTC (67ab5d7)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c-ccg/vc-api/254/1714ec7...67ab5d7.html" title="Last updated on Jan 7, 2022, 4:53 PM UTC (67ab5d7)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -493,6 +493,240 @@ <h4>Update Status</h4>
...
+        <h4>Automation via Workflows</h4>
+
+        <p class="issue" title="EXPERIMENTAL FEATURE">
+Workflows are an experimental feature that is currently being explored by the
+group. This feature is likely to undergo significant revision as conversations
+around the feature progress.
+        </p>
+
+        <p>
+The APIs in this specification enables automated (machine-to-machine) or
```
Comment (TallTed) [inline index.html:506]:
```suggestion
The APIs in this specification enable automated (machine-to-machine) or
```

File: index.html [hunk 2]
```diff
@@ -493,6 +493,240 @@ <h4>Update Status</h4>
...
+        </p>
+
+        <ol>
+          <li>
+The Holder App contacts the Issuer App to request the initiation of a particular
+workflow.
+          </li>
+          <li>
+The Issuer App responds with a presentation request of some kind to authenticate
+and/or authorize the Holder App and provides the next hop in the workflow as  a
```
Comment (TallTed) [inline index.html:519]:
```suggestion
and/or authorize the Holder App, and provides the next hop in the workflow as a
```

File: index.html [hunk 3]
```diff
@@ -493,6 +493,240 @@ <h4>Update Status</h4>
...
+      }
+    }
+  }
+}
+        </pre>
+
+        <p class="note" title="Interact service description">
+Readers are urged to pay particular attention to the `interact` service
+description that provides a mechanism to switch between automated flows and
+ones requiring mediation by a person.
```
Comment (TallTed) [inline index.html:594]:
```suggestion
flows requiring mediation by a person.
```

File: index.html [hunk 4]
```diff
@@ -493,6 +493,240 @@ <h4>Update Status</h4>
...
+        </p>
+
+        <p class="note">
+While the example above utilizes a Verifiable Presentation Request, other types
+of equivalent presentation request formats, such as WACI/PeX, can be utilized
+(simultaneously or alternatively) based on the presentation request formats that
+are supported by the Issuer. Automated requests can be signalled by setting  the
+service `type` to `UnmediatedPresentationService2021`, while manual  requests
```
Comment (TallTed) [inline index.html:602]:
```suggestion
are supported by the Issuer. Automated requests can be signaled by setting the
service `type` to `UnmediatedPresentationService2021`, while manual requests
```

File: index.html [hunk 5]
```diff
@@ -493,6 +493,240 @@ <h4>Update Status</h4>
...
+can be signalled by setting the service `type` to
```
Comment (TallTed) [inline index.html:603]:
```suggestion
can be signaled by setting the service `type` to
```

**Review by OR13 — CHANGES_REQUESTED** [review]
I like this idea, but its sorta already supported by the existing endpoints... and I question some of the redundancy regarding "session" or "flow identifiers"... looking for a review from @mavarley @peacekeeper @mprorock and @mkhraisha before I approve.

File: index.html [hunk 6]
```diff
@@ -493,6 +493,240 @@ <h4>Update Status</h4>
...
+workflow.
+          </li>
+          <li>
+The Issuer App responds with a presentation request of some kind to authenticate
+and/or authorize the Holder App and provides the next hop in the workflow as  a
+URL.
+          </li>
+          <li>
+The Holder App responds to the Issuer App with a Verifiable Presentation
+containing information that will satisfy the presentation request.
```
[the remark this answers is not in the document]
Comment (msporny) [reply index.html:524]: Well, the question that this PR raises is "do we want to have the option to enable other request formats?". I'm suggesting that we do in order to head off a "presentation request" argument. The downside here is that it's variability. It's a trade-off... but that's why we use `verifiablePresentationRequest`, to provide that extension point.

File: index.html [hunk 7]
```diff
@@ -493,6 +493,240 @@ <h4>Update Status</h4>
...
+            "type": "UniversityDegreeCredential"
+          }
+        }
+      }],
+      "challenge": "3182bdea-63d9-11ea-b6de-3b7c1404d57f",
+      "domain": "example.edu",
+      "interact": {
+        "service": [{
+          "type": "UnmediatedPresentationService2021",
+          "serviceEndpoint": "https://example.edu/active-flows/123456"
```
[the remark this answers is not in the document]
Comment (msporny) [reply index.html:583]: Is that conversation public where I could read up on it?

File: index.html [hunk 8]
```diff
@@ -493,6 +493,240 @@ <h4>Update Status</h4>
...
+of equivalent presentation request formats, such as WACI/PeX, can be utilized
+(simultaneously or alternatively) based on the presentation request formats that
+are supported by the Issuer. Automated requests can be signalled by setting  the
+service `type` to `UnmediatedPresentationService2021`, while manual  requests
+can be signalled by setting the service `type` to
+`MediatedPresentationService2021`.
+        </p>
+
+        <p class="issue" title="Demonstrate usage of WACI/PeX?">
+We should make it clear that Issuers can choose to use one or more presentation
```
[the remark this answers is not in the document]
Comment (msporny) [reply index.html:608]: Yeah, that's why I left this as an issue. Do we want to run DIDComm over VC API (does that even make sense?). Do we want to run WACI/PeX over VC API (I think that makes more sense than DID Comm)? I expect these to be topics of conversation when we discuss this in next week's VC API call.

[the remark this answers is not in the document]
Comment (OR13) [reply index.html:608]: I would consider DIDComm and PEx / WACI out of scope for this work item until there is a resolution to move them in scope. I would object to moving them in scope at this time.

File: index.html [hunk 9]
```diff
@@ -493,6 +493,240 @@ <h4>Update Status</h4>
...
+request formats simultaneously in order to maximize the chances that a given
+client will be able to respond to a refresh workflow. This should probably be
+more clearly stated in the VC-API specification.
+        </p>
+
+        <pre class="example"
+          title="Step 3 (presentation to issuer): Provide Verifiable Presentation">
+POST /active-flows/123456 HTTP/1.1
```
[the remark this answers is not in the document]
Comment (OR13) [reply index.html:616]: In our existing work in trace interop, and implementer always has a known endpoint for receiving presentations... and that endpoint is already defined here:

https://github.com/w3c-ccg/vc-api/blob/main/holder.yml#L59

perhaps this first example added could use the existing endpoints? (domain and challenge are passed via post body today, not via URL param... this is better when considering non http transports).

**Comment by OR13** [comment]
@msporny we've demonstrated interop with the endpoints in the yaml files that in currently on main, specifically:

- https://github.com/w3c-ccg/vc-api/blob/main/holder.yml#L37
- https://github.com/w3c-ccg/vc-api/blob/main/holder.yml#L59

The exchange protocol we use is already the vp-request-spec.

here is an open source implementation of them:

https://api.did.actor/docs#post-/api/presentations/available
https://api.did.actor/docs#post-/api/presentations/submissions

Transmute, Mesur and Mavennet have used these endpoints for some months now, to exchange presentations. 

We even have a credential which announces that an organization supports them:

https://w3id.org/traceability/#VerifiableBusinessCard

^ this is similar to the proposed work around for "DID should not have service endpoints which reveal information".

**Comment by msporny** [comment]
> TLDR, these changes seem to add valuable documentation for features that are already in use.... and is very much welcome.

Yeah, so maybe this is either 1) documentation about the existing endpoints, or 2) minor modifications of the existing endpoints, or 3) major modifications to the existing endpoints. 

Here are a few of the concerns w/ the existing endpoints as documented:

* The language around "presentations/available" and "presentations/submissions" are confusing to the point where at least I continue to be confused around the overlap between this PR and those endpoints. Every time I think I understand what they're doing, something pops up that feels new or missing. So, at the very least, we need to have a bikeshedding discussion around those endpoints, if (in fact) they are exactly the same as the endpoints described in this email.
* It is not clear at all that those endpoints are meant for multi-hop interactions.
* You linked to https://github.com/w3c-ccg/vc-api/blob/main/holder.yml#L37 above -- this is for "/presentations/prove", which just attaches a proof to a presentation, right? While a useful feature, that is out of scope for this PR (while there is an assumption that that is happening behind the scenes, nothing in the workflows above do that).
* You linked to https://github.com/w3c-ccg/vc-api/blob/main/holder.yml#L59 above -- this is for `/presentations/available`, and maps to steps 1 and 2 here https://pr-preview.s3.amazonaws.com/w3c-ccg/vc-api/pull/254.html#automation-via-workflows -- but without the interaction or presentation request property, which is important for a variety of reasons: 1) to distinguish between mediated and unmediated flows, 2) to provide a next-hop that isn't hard coded to `/presentations/submissions`, which is important for decentralized workflows (pick up VCs from multiple parties), and 3) to support other presentation request languages.
* I believe step 3 and 4 of the automation flow could loosely be mapped to `/presentations/submissions`, except that there is 1) no `interaction` continuation possibility (which defeats the purpose of workflows), and 2) there is a presumption that there is only one endpoint that can be used to submit presentations (and that's a bad choice for some stateful load balancing and/or security architectures). The endpoint is described as "Provide a presentation for a holder to store." which is part of the thing that's confusing because it indicates that an issuer couldn't use this as part of an issuance workflow (as this PR does). This is probably just bad wording/explanation in the existing spec text?

I have been wanting all of us as a group to have a conversation around the (confusingly named) "Holder APIs" in the yaml (called Presentation APIs in the spec). Given that all of this is required to be resolved in order to get the Verifiable Credential Refresh 2021 spec done and test suited, we might want to prioritize this conversation in the next VC API call.

Thoughts?

**Comment by OR13** [comment]
@msporny yes, we have song since distinguish APIs for exchanging presentations from APIs for creating presentations:

(Derive and Prove)
- https://w3id.org/traceability/interoperability#presentation

(Notify of workflow, and submit presentation)
- https://w3id.org/traceability/interoperability#exchange

These APIs don't support multi hop flows... today, they support Holder -> Verifier flows over http.

`interaction` is defined where exactly? IIRC @dmitrizagidulin invented it based on GNAP.

I think its potentially interesting but it should define here: https://w3c-ccg.github.io/vp-request-spec/

Before being referenced from this repo.

File: index.html (hunk 7 above)
[the remark this answers is not in the document]
Comment (OR13) [reply index.html:583]: related PEx concept: https://identity.foundation/waci-presentation-exchange/#web-redirect

There was a demo of this given at a meeting a while back.

There are examples which are mostly related to DIDComm and WACI / PEx here: https://w3c-ccg.github.io/universal-wallet-interop-spec/#examples

Some of these examples seem slightly out of date with latest versions of PEx.

The main point was that SecureKey had looked at presentation exchange via web mediators. 

I can't find the best source, which Is why I asked fro them to comment.

File: index.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (OR13) [reply index.html:524]: I don't think we have a resolution that allows us to consider other request formats. 

I recall we resolved to do our best to use vp-request spec. 

I am ok defining these APIs in a way that doesn't preclude other presentation requests formats such as PEx / WACI.... but I think its a mistake to try and support N community drafts that are being updated daily at this point.

Let's pick a single first way to do things and make it work before adding "alternatives" that will create interoperability issues.... (lets avoid the request spec negotiation argument for as long as we can).

**Comment by peacekeeper** [comment]
My reaction was the same as Orie's.. Looks very interesting, but I'm wondering about some details of the flow and whether it can better re-use some existing web and auth concepts. For example:
- Maybe step 2 should return 401 instead of 200.
- Maybe steps 3 and 4 are a bit like accessing an OAuth2 protected resource, and an access token should be issued, or a cookie set, or something like that.

This kind of interaction over multiple steps is supposed to be one of the strengths of DIDComm and its (sub-)protocols. But I'd also be fine with keeping it simple and doing it as proposed here in this PR.

If I understand correctly, these workflows are meant to be a more general concept, not limited to the re-issuing of credentials use case?

Another hip term I see more and more recently is "playbook" rather than "workflow".

**Comment by OR13** [comment]
@peacekeeper @msporny perhaps this PR here helps explain the current API definitions for presentation exchange that are committed to main in this repo.

https://github.com/w3c-ccg/vp-request-spec/pull/12

I feel like we might be doing this "request query model definition" in the wrong repo... This repo is supposed to just define how to use vp-request spec, and vc data model over http... we should not be defining vp-request spec data models in this work item, imo.

**Comment by msporny** [comment]
@OR13 wrote:
> @msporny yes, we have tried to distinguish APIs for exchanging presentations from APIs for creating presentations:

Yes, though since they're all bunched under "/presentations/*" -- it confuses things a bit. One thing we could try is renaming:

`/presentations/available` to `/presentations/start`

... that is, if we want a static starting point that all verifiers accept (which seems like a reasonable thing to have). However, we don't want to limit it to just that static endpoint (for the load balancing and architectural reasons I mentioned before). We want to establish the `/presentations/start` as a default and a pattern (and not just a single endpoint), IMHO.

So, the first question is, would there be an objection to renaming `/presentations/available` to `/presentations/start`. I'm objecting to the `/presentations/available` endpoint as being misleading (this is one of the things that created months of confusion for me wrt. the purpose of the endpoint).

> These APIs don't support multi hop flows... today, they support Holder -> Verifier flows over http.
> 
> `/presentations/available` maps to the logical concept of:
> 
> "As a holder, I want to present to you credentials, please give me a domain and challenge".

Yes, agreed that we need that concept to be realized... we're just bike-shedding the name at this point.

> We stopped working on "workflows" and "allow lists" after we realized that the channel would probably be authenticated via Client Credentials, at least for machine to machine connections.

Right, and I'm asserting that assuming that there will be authentication via Client Credentials (assuming you mean OAuth2, here?) is not something that's universal. There are use cases where you don't need any Client Credentials, or the URL endpoint can act as a credential in itself. In other words, there are multiple ways to address this and forcing one way seems like the wrong way to go here (OAuth2 Client Credentials create a burden that won't work for our at least one of our big credential refresh flows, for example). How to secure each leg of the flow should be left up to the implementation.

> We also determined that "workflows" can be considered presentations of credentials with at least 1 unique correlation identifier, and then a holder -> verifier presentation is sufficient.

Sure, that seems like a fine working theory that'll hold. We just need to make sure that we're not imposing the OAuth2 Client Credentials as a requirement when it's not necessary for some flows.

> `interaction` is defined where exactly? IIRC @dmitrizagidulin invented it based on GNAP.
> I think its potentially interesting but it should define here: https://w3c-ccg.github.io/vp-request-spec/

Agreed that we should define it in vp-request-spec. I was only trying to highlight how it's used in practice in the credential refresh spec as well as this spec as illustration. I'll add a PR to VPR that adds `interact`.

**Comment by msporny** [comment]
@peacekeeper wrote:
> * Maybe step 2 should return 401 instead of 200.

Yes, good catch @peacekeeper. We should certainly keep the option of responding with a 401 possible. In this particular case, a 401 isn't necessary because the example contemplates an open endpoint where you need to DIDAuth to in step 3.

I do agree, however, that we should make sure people know that a 401 is possible to do after step 1. We also have to be careful to make sure it's not mandatory (because there are other ways of authenticating -- like DIDAuth -- that don't use HTTP authn mechanisms).

> * Maybe steps 3 and 4 are a bit like accessing an OAuth2 protected resource, and an access token should be issued, or a cookie set, or something like that.

Yes, Cookies or OAuth2 Client Credentials could be one way to securing steps 3 and 4. I will also note that the `interact` mechanism in step 2 can provide an authentication token (as a parameter in the URL). So, while we want to enable existing authn mechanisms for steps 3 and 4, we don't want to mandate them (as simpler authn mechanisms might be more appropriate).

> This kind of interaction over multiple steps is supposed to be one of the strengths of DIDComm and its (sub-)protocols. But I'd also be fine with keeping it simple and doing it as proposed here in this PR.

One of the options is for step 2 to provide a DIDComm communication channel, either via the `interact` mechanism (a DIDComm endpoint), or via a separate presentation request parameter that is parallel to `verifiablePresentationRequest`.

> If I understand correctly, these workflows are meant to be a more general concept, not limited to the re-issuing of credentials use case?

Yes, exactly, they are meant to be a generalized workflow mechanism for Holders to request certain workflows that will result in the exchange of one or more VPR and VPs, resulting in a set of VCs being issued throughout the process or just an acknowledgement that the VPs have been submitted to the verifier. If we do this correctly, this simple four call flow will be able to handle 80% or more of the VP exchange in the ecosystem (and handle multiple presentation protocols if needed).

> Another hip term I see more and more recently is "playbook" rather than "workflow".

As a mostly native English speaker, "playbook" feels less accurate than workflow... where you start in one place and you go through a series of steps to reach an end result. A playbook is more of a series of strategies, though I do see where it's hard to distinguish one from the other.

**Comment by msporny** [comment]
> I feel like we might be doing this "request query model definition" in the wrong repo... The vc-api repo is supposed to just define how to use vp-request spec, and vc data model over http... we should not be defining vp-request spec data models in this work item, imo.

Yeah, this stuff probably needs to go in multiple repos, though we do have to define at least one thing here, which is the protocol object in step 2.

While I don't want to necessarily get into this discussion yet... if you look at these four calls as a design pattern more than a concrete set of APIs, you might be able to see how you could start a workflow for issuing an ISO-18013-5 mDL (as in a bonafide one, not a Verifiable Driver's License alternative). You might also see how you could start a workflow where the Issuer/Verifier gives you the option of continuing using VPR or WACI/PeX. While it's far simpler to just have one protocol with one presentation request format and one response format... it is possible with this design pattern to act as convergence mechanism for a variety of the presentation request formats (VPR, PeX) as well as existing credential formats (VC, mDL, AnonCreds) that exist today.

I suggest we get the VC API / VPR stuff right first... but we should keep the convergence story in mind as we keep building (because it would be a powerful unifier of multiple communities if we can pull it off).

**Comment by msporny** [comment]
I have done as @OR13 requested and put in a PR to the VPR spec to add the `interact` paradigm there: https://github.com/w3c-ccg/vp-request-spec/pull/13

**Review by OR13 — APPROVED** [review]

**Comment by OR13** [comment]
This PR does not preclude the idea that an `interact` service would be an instance of the `vc-api`... a future PR should provide an example where that is the case, or clear language saying its forbidden.

**Review by mprorock — APPROVED** [review]
Review on VC-API call helped quite a bit.

**Comment by msporny** [comment]
Multiple reviews, changes requested and made, no objections, merging. 

The next PR will build on advice given during the VC API call to either re-use existing endpoints or to define new concrete endpoints that realize this flow. Thanks, everyone, for the engagement on this PR, more to come, soon!
